xarray/core/indexing.py

```diff
@@ -1433,19 +1484,31 @@ def __getitem__(self, key):
         array, key = self._indexing_array_and_key(key)
         return array[key]
 
-    def __setitem__(self, key, value):
-        array, key = self._indexing_array_and_key(key)
+    def _safe_setitem(self, array, key, value):
         try:
             array[key] = value
-        except ValueError:
+        except ValueError as exc:
             # More informative exception if read-only view
             if not array.flags.writeable and not array.flags.owndata:
                 raise ValueError(
                     "Assignment destination is a view.  "
                     "Do you want to .copy() array first?"
                 )
             else:
-                raise
+                raise exc
+
+    def _oindex_set(self, key, value):
+        key = _outer_to_numpy_indexer(key, self.array.shape)
+        self._safe_setitem(self.array, key, value)
+
+    def _vindex_set(self, key, value):
+        array = NumpyVIndexAdapter(self.array)
+        self._safe_setitem(array, key.tuple, value)
+
+    def __setitem__(self, key, value):
+        self._check_and_raise_if_non_basic_indexer(key)
+        array, key = self._indexing_array_and_key(key)
```

Comment:
I believe we can extract the one clause we need from `_indexing_array_and_key`, add it to both `__getitem__` and `__setitem__` and delete `_indexing_array_and_key`.